core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/SegmentIndexFileStore.java

```diff
@@ -108,11 +109,20 @@ public void readAllIIndexOfSegment(SegmentFileStore segmentFileStore, SegmentSta
           location =
               segmentFileStore.getTablePath() + CarbonCommonConstants.FILE_SEPARATOR + location;
         }
+        String mergeFileName = locations.getValue().getMergeFileName();
         for (String indexFile : locations.getValue().getFiles()) {
           CarbonFile carbonFile = FileFactory
               .getCarbonFile(location + CarbonCommonConstants.FILE_SEPARATOR + indexFile);
-          if (carbonFile.exists()) {
+          if (carbonFile.exists() && !indexFiles.contains(carbonFile.getAbsolutePath())) {
             carbonIndexFiles.add(carbonFile);
+            indexFiles.add(carbonFile.getAbsolutePath());
+          } else if (mergeFileName != null) {
```

Comment:
Can move mergeFilename logic out of loop

Reply:
ok